### PR TITLE
Scrm 585 flag btn

### PR DESCRIPTION
**Pull request by gymad**
## Description

Adding: 
- mark as flag buttons to bulk action onto email list view
- tidy up indicators style
- fix uids at checkboxes
(ref: SCRM-585, #3610 )

## Motivation and Context

Mark as Flag/Unflagged/Read/Unread function are missing

## How To Test This

- rebuild style.css (sass)
- QRR
- check emails list view => bulk action => "marks as ..." menuitems

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [ ] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ ] My change requires a change to the documentation.
- [ ] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: modules/Emails/include/ListView/ImportEmailAction.js [hunk 1]
```diff
@@ -56,53 +56,16 @@
...
-        } else {
-          console.error('Error importing emails. Please check the logs for details.')
-        }
-      }).error(function (data) {
-        mb.hide();
-        alert(data);
-      });
+        },
+        false,
+        SUGAR.language.translate('Emails', 'LBL_IMPORTING'),
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ImportEmailAction.js]: Why not translate the entire message?

File: modules/Emails/include/ListView/ListViewHeader.js [hunk 2]
```diff
@@ -37,18 +37,88 @@
...
+ * @param {string} actionUrl
+ * @param {function} successCallback
+ * @param {function} errorCallback
+ * @param {string} loadingTitle
+ * @param {string} errorMessage
+ */
+var handleSelectedEmailsListViewItems = function(actionUrl, successCallback, errorCallback, loadingTitle, errorMessage) {
+
+  if(typeof loadingTitle === 'undefined') {
+    SUGAR.language.translate('Emails', 'LBL_LOADING');
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewHeader.js]: Translation does not do anything. Where is the assignment?

File: modules/Emails/include/ListView/ListViewHeader.js [hunk 3]
```diff
@@ -37,18 +37,88 @@
...
+  }
+
+  if(typeof errorMessage === 'undefined') {
+    errorMessage = 'Error at selected emails handling.';
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewHeader.js]: Why translate one error and not this error?

File: modules/Emails/include/ListView/ListViewHeader.js [hunk 4]
```diff
@@ -37,18 +37,88 @@
...
+};
+
 $(document).ready(function () {
 
   var query = JSON.parse($('[name=current_query_by_page]').val());
+  var $btnEmailsCurrentFolder = $('.btn-emails-current-folder');
+
   if(typeof query.folder === 'undefined' ||  query.folder === '') {
-    $('.btn-emails-current-folder').remove();
+    $btnEmailsCurrentFolder.remove();
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewHeader.js]: Missing () perhaps. Please do not use $variable for varable names. The $ sign signifies dependency injection in MVVM frameworks like angular. Plus it looks like broken jQuery.

File: modules/Emails/include/ListView/MarkEmails.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      "use strict";
+
+      handleSelectedEmailsListViewItems(
+        'MarkEmails&type=' + $(this).attr('data-for'),
+        function() {
+          window.location.reload();
+        },
+        false,
+        SUGAR.language.translate('Emails', 'LBL_MARKING'),
+        'Error marking emails.'
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/MarkEmails.js]: Again not translated?

File: modules/Emails/include/ListView/MarkEmails.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+(function ($) {
+  /**
+   *
+   * @param options
+   * @return {*|HTMLElement}
+   */
+  $.fn.MarkEmails =  function(options) {
+    "use strict";
+    var self = {};
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/MarkEmails.js]: please change self = {} to self = this

File: modules/Emails/include/ListView/MarkEmails.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    self.construct = function () {
+      "use strict";
+      $(opts.buttonSelector).click(self.handleClick);
+    };
+
+    /**
+     * @destructor
+     */
+    self.destruct = function() {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/MarkEmails.js]: The destruct is not being setup in the constructor. You could add in the self.construct method when this has been removed:

`$(this).on('removed', self.destruct);`

It is a good idea to raise an event in the constructor and destructor.

File: modules/Emails/include/ListView/MarkEmails.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    };
+
+    self.construct();
+    return $(this);
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/MarkEmails.js]: change to return $(self);

File: modules/Emails/include/ListView/ListViewHeader.js [hunk 9]
```diff
@@ -37,18 +37,88 @@
...
 
+/**
+ * @todo take this function to a common email listview object
+ * @param {string} actionUrl
+ * @param {function} successCallback
+ * @param {function} errorCallback
+ * @param {string} loadingTitle
+ * @param {string} errorMessage
+ */
+var handleSelectedEmailsListViewItems = function(actionUrl, successCallback, errorCallback, loadingTitle, errorMessage) {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewHeader.js]: Should we put this into a jQuery plugin? Perhaps in a simaler fashion to click() or focus() ?

File: include/MVC/View/views/view.list.php [hunk 10]
```diff
@@ -128,7 +128,7 @@ public function listViewPrepare()
         $metadataFile = $this->getMetaDataFile();
 
         if (!file_exists($metadataFile))
-            sugar_die($GLOBALS['app_strings']['LBL_NO_ACTION']);
+            sugar_die(sprintf($GLOBALS['app_strings']['LBL_NO_ACTION'], $this->do_action));
```
Comment (daniel-samson) [inline include/MVC/View/views/view.list.php]: PSR2: please add the {} to the if statement

File: modules/Emails/controller.php [hunk 11]
```diff
@@ -442,4 +446,92 @@ public function composeBean($request, $mode = self::COMPOSE_BEAN_MODE_UNDEFINED)
...
+        $request = $_REQUEST;
+
+        // validate the request
+
+        if(!isset($request['inbound_email_record']) || !$request['inbound_email_record']) {
+            throw new SugarControllerException('No Inbound Email record in request');
+        }
+
+        if(!isset($request['folder']) || !$request['folder']) {
+            throw new SugarControllerException('No Inbound Email folder in request');
```
Comment (daniel-samson) [inline modules/Emails/controller.php]: SugarControllerException? Where is it?

File: modules/Emails/controller.php [hunk 12]
```diff
@@ -442,4 +446,92 @@ public function composeBean($request, $mode = self::COMPOSE_BEAN_MODE_UNDEFINED)
...
+
+        $UIDs = $this->getRequestedUIDs($request);
+        $type = $this->getRequestedFlagType($request);
+
+        // mark emails
+
+        $ie->markEmails($UIDs, $type);
+
+        echo json_encode(array('response' => true));
+        die();
```
Comment (daniel-samson) [inline modules/Emails/controller.php]: don't use die. Instead set the view to ajax:

`$this->view = 'ajax'`

File: modules/Emails/controller.php [hunk 13]
```diff
@@ -442,4 +446,92 @@ public function composeBean($request, $mode = self::COMPOSE_BEAN_MODE_UNDEFINED)
...
+     * @param $request
+     * @return null|mixed
+     */
+    private function getRequestedFlagType($request) {
+        $ret = $this->getRequestedArgument($request, 'type');
+        return $ret;
+    }
+
+    /**
+     * @param $request
```
Comment (daniel-samson) [inline modules/Emails/controller.php]: please update php doc so that it specifies the param type

File: modules/Emails/controller.php [hunk 14]
```diff
@@ -442,4 +446,92 @@ public function composeBean($request, $mode = self::COMPOSE_BEAN_MODE_UNDEFINED)
...
+            $GLOBALS['log']->error("Requested key is not set: ");
+            return null;
+        }
+        return $request[$key];
+    }
+
+    /**
+     * return an Inbound Email by requested record
+     *
+     * @param $record
```
Comment (daniel-samson) [inline modules/Emails/controller.php]: Again please update php doc

File: modules/Emails/include/ListView/ImportEmailAction.js (hunk 1 above)
Comment (gymad) [inline modules/Emails/include/ListView/ImportEmailAction.js]: good question, legacy change from your commit: 2e1cb22. the errorMessage passed into a console.error(...) which should not translated, however the message box shows directly to end-user, who may  don't speak English.

File: modules/Emails/include/ListView/ListViewHeader.js (hunk 3 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewHeader.js]: see above.

File: modules/Emails/include/ListView/MarkEmails.js (hunk 6 above)
Comment (gymad) [inline modules/Emails/include/ListView/MarkEmails.js]: similar issues in:
include/javascript/EmailsComposeViewModal.js
jssource/src_files/include/javascript/EmailsComposeViewModal.js
modules/Emails/include/ListView/CheckNewEmails.js
modules/Emails/include/ListView/FoldersViewModal.js
modules/Emails/include/ListView/SettingsView.js

File: modules/Emails/include/ListView/ListViewHeader.js (hunk 4 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewHeader.js]: ok

File: modules/Emails/include/ListView/MarkEmails.js (hunk 5 above)
Comment (gymad) [inline modules/Emails/include/ListView/MarkEmails.js]: again see above

File: modules/Emails/include/ListView/MarkEmails.js (hunk 8 above)
Comment (gymad) [inline modules/Emails/include/ListView/MarkEmails.js]: similar issues:
include/javascript/EmailsComposeViewModal.js
jssource/src_files/include/javascript/EmailsComposeViewModal.js
modules/Emails/include/ListView/CheckNewEmails.js
modules/Emails/include/ListView/SettingsView.js

File: modules/Emails/controller.php (hunk 11 above)
Comment (gymad) [inline modules/Emails/controller.php]: added

File: modules/Emails/controller.php (hunk 12 above)
Comment (gymad) [inline modules/Emails/controller.php]: everything broken if you do this

**Pull request merged**